### PR TITLE
[persist] Fast-path peek prototype

### PR DESCRIPTION
**Pull request by bkirwi**
This introduces a new type of fast-path peek, which will read data from Persist directly instead of spinning up a dataflow.

### Motivation

See #20708 for the design and https://github.com/MaterializeInc/materialize/issues/21265 for the epic.

### Tips for reviewer

~I haven't benchmarked this properly yet, but the overall results seem promising: in testing I'm often seeing roughly an order-of-magnitude improvement on my little staging cluster. (For example: 2s vs 16s for a 4m-row table.)~ I've done a bit more testing: the latency seems to be dominated by the slowest part fetch plus the cost of sorting individual parts, since my test dataset (perhaps unrealistically) includes some large unconsolidated parts. I feel like we could probably drive this latency down further, but mostly with changes unrelated to this PR, and the speedup is already pretty good....

One interesting thing to note is that most of the savings seem to be due to serde and not to less data being fetched, since only massive shards tend to have more than one part per run. Leverage our `key_lower` stats, which lets us skip some runs entirely, may push this down further.

Known limitations:
- These fast-path peeks are not cancellable.
- There is no global limit on the number of fast-path peeks that can be running at once, and thus no limit on memory use.

For this reason I don't want to enable the flag right away, but it seems like I ought to have it enabled somewhere. So I've enabled it in testdrive only. And it's possible we decide both these things are fine?

A version of this branch with the fast-path unconditionally enabled passes nightlies, which is good to see.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-client/src/controller.rs
Comment (bkirwi) [reply src/storage-client/src/controller.rs:1799]: Huh, yeah, this is wrong! Will think it over.

Comment (bkirwi) [reply src/storage-client/src/controller.rs:1814]: Tried both: 2 didn't work out nicely - the straightforward approach made the controller not object-safe - and 1 was better than I expected, so went with that for now.

File: src/persist-client/src/internal/metrics.rs
Comment (bkirwi) [reply src/persist-client/src/internal/metrics.rs:2005]: I'm passionate about this histogram, because I was referring to these stats even during testing.

For the flag in this PR we'll only end up running this code in this envd, but if we eg. roll out the streaming compaction flag we'd end up doing this all over the place.

File: src/persist-client/src/read.rs
Comment (bkirwi) [reply src/persist-client/src/read.rs:1061]: It turns out I can fall back to the derived debug impl here - the explicit impl was a bit of an artifact of the box-future stuff from the last PR, so I'l do that!

File: src/persist-client/src/internal/metrics.rs
Comment (danhhz) [reply src/persist-client/src/internal/metrics.rs:2005]: discussed this offline. I'm on board with a histogram for the total latency for executing an entire fast path peek. I'm less sold on needing one for join latency, but my objection is more "in principle" than "in practice", so if you end up feeling strongly on two histograms, that's fine

File: src/storage-client/src/controller.rs
Comment (danhhz) [inline src/storage-client/src/controller.rs:189]: let's include a comment on why we did this

**Review by danhhz — APPROVED** [review]
everything but src/adaptor changes lgtm! let's pull in an expert to review those bits

File: src/adapter/src/metrics.rs
Comment (bkirwi) [inline src/adapter/src/metrics.rs:30]: I'm not sure how adapter likes to structure metrics... happy to take advice here!

**Comment by bkirwi** [comment]
Alright, I've fixed the first round of issues... opening this up for wider review.

As context for folks: this is meant to be a first pass at #21615. All the changes here ought to be protected by the new flag, and we don't have plans to ramp that flag up right away in prod. Please do mention if there's anything you'd like to see before this goes to production, though I will likely ask if I can take anything nontrivial as followup work.

File: src/adapter/src/coord/peek.rs
Comment (jkosh44) [inline src/adapter/src/coord/peek.rs:570]: I think we should add a new execution strategy for this. The description of `StatementExecutionStrategy::FastPath` is 

> The statement was executed by reading from an existing arrangement.

which isn't exactly accurate here.

File: src/adapter/src/coord/peek.rs
Comment (jkosh44) [inline src/adapter/src/coord/peek.rs:584]: We generally ignore these errors and log a warning. They generally only happen when the Coordinator thread has shut down before this task. For example:

https://github.com/MaterializeInc/materialize/blob/b5e366cea8acb94e95b808db2ab3918cfb77cfc2/src/adapter/src/coord.rs#L879-L884

**Review by jkosh44 — COMMENTED** [review]
Can you add some tests that execute what we would expect are fast path persist peeks and assert that they actually are? You can probably just look at EXPLAIN outputs for certain `SELECT`s.

File: src/adapter/src/metrics.rs
Comment (jkosh44) [reply src/adapter/src/metrics.rs:30]: https://materializeinc.slack.com/archives/C02FWJ94HME/p1694805197123109 I asked in the surfaces channel, we've just recently been thinking about questions like this, but I'm not sure if we have a concrete answer yet.

File: src/adapter/src/metrics.rs
Comment (mjibson) [inline src/adapter/src/metrics.rs:93]: This text is wrong.

File: src/adapter/src/metrics.rs
Comment (mjibson) [reply src/adapter/src/metrics.rs:30]: We don't have many opinions yet. I think it's fine as written here and we can refactor if desired?

File: src/adapter/src/coord/peek.rs
Comment (bkirwi) [reply src/adapter/src/coord/peek.rs:309]: This is a copy-paste of some existing mfp-to-safemfp conversions in the file, including `permute_oneshot_mfp_around_index` which is invoked just below. I don't have any additional context.

Maybe I should factor this out to a function so that it's easier for someone to add a more structured error in the future?

File: src/adapter/src/coord/peek.rs
Comment (bkirwi) [reply src/adapter/src/coord/peek.rs:564]: If the task gets cancelled. (In theory we'd get panic errors here too, but we abort on panic.) This shouldn't happen in the existing code much, since we don't have proper cancellation, but I figured I might as well wire it up. (IIRC it may also happen if the async runtime is being shut down, but I think we don't care that much about what happens in that case.)

File: src/adapter/src/coord/peek.rs
Comment (jkosh44) [reply src/adapter/src/coord/peek.rs:309]: > Maybe I should factor this out to a function so that it's easier for someone to add a more structured error in the future?

I think that's fine for now.

File: src/adapter/src/coord/peek.rs
Comment (bkirwi) [reply src/adapter/src/coord/peek.rs:536]: Yes, running in environmentd!

Is it okay if I take this as a followup? I would like to roll this change in with some of the other limit-enforcement stuff mentioned in the PR description to make it easier to test. In the meantime this will stay guarded behind the flag.

Comment (jkosh44) [reply src/adapter/src/coord/peek.rs:536]: Sure, would you mind adding a comment or issue so it doesn't slip through the cracks?

**Review by jkosh44 — APPROVED** [review]

**Comment by bkirwi** [comment]
Thanks! I'll hold off on merging until tomorrow to let nightlies run and in case anyone else wants to 👁️:eye:.

File: src/adapter/src/coord/peek.rs
Comment (bkirwi) [reply src/adapter/src/coord/peek.rs:536]: Certainly: added to the task list at https://github.com/MaterializeInc/materialize/issues/21265

**Comment by philip-stoev** [comment]
I would like to test this. Please hold off merging until I sign off. Thanks!

File: test/testdrive/mzcompose.py
Comment (philip-stoev) [inline test/testdrive/mzcompose.py:30]: This does not belong here. It is best to put it in `misc/python/materialize/mzcompose/__init__.py` , `DEFAULT_SYSTEM_PARAMETERS` , so that it applies to all mzcompose-based tests, including e.g. Zippy and the various random query generators. The default should ideally be the same as the one used in production.

File: test/sqllogictest/persist-fast-path.slt
Comment (philip-stoev) [inline test/sqllogictest/persist-fast-path.slt:81]: Thank you for the tests. A test with `OFFSET` would be nice to have -- the optimization currently does not kick in, even though theoretically it could:

```
materialize=> EXPLAIN SELECT * from events limit 1 offset 2;
               Optimized Plan               
--------------------------------------------
 Explained Query:                          +
   Finish limit=1 offset=2 output=[#0..=#2]+
     ReadStorage materialize.public.events +
 
(1 row)
```

**Comment by philip-stoev** [comment]
Item No 1 With a `generate_series(1,10 * 1000 * 1000)` dataset, the performance is identical with the new path and the old one:

```
materialize=> select * from t1 limit 10000;
Time: 4807.287 ms (00:04.807)
materialize=> select * from t1 limit 1;
 f1 
----
  1
(1 row)

Time: 4033.245 ms (00:04.033)
```

Does that match your observations? I also tried a table where all values are identical and got identical results.

**Review by philip-stoev — CHANGES_REQUESTED** [review]
Thank you. Seems to work as advertised, albeit with negligible performance improvement.
I am setting this to "Request changes" because of there are no non-EXPLAIN tests for the feature.

File: test/sqllogictest/persist-fast-path.slt
Comment (philip-stoev) [inline test/sqllogictest/persist-fast-path.slt:30]: A non-`EXPLAIN` query is needed here that confirms that correct data is actually being returned for eligible queries.

File: test/testdrive/mzcompose.py
Comment (bkirwi) [reply test/testdrive/mzcompose.py:30]: I wrote a bit about the rationale for this in the PR description.

At the moment I would prefer to leave it disabled entirely in CI than to have it enabled everywhere, because we have not yet done the hardening or testing that's required to be comfortable turning this on in prod. Let me know if you'd like that better, or if we should find some other way to get a small amount of coverage in the meantime.

**Comment by bkirwi** [comment]
> Does that match your observations?

It does not! A quick local run:

```
materialize=> insert into philip select * from generate_series(1,10 * 1000 * 10);
INSERT 0 100000
materialize=> insert into philip select * from generate_series(1,10 * 1000 * 100); -- smaller limits because i'm running in dev mode
INSERT 0 1000000
Time: 64217.037 ms (01:04.217)
materialize=> select * from philip limit 10; -- with `ALTER SYSTEM SET persist_fast_path_limit=1000;`
 foo 
-----
   1
   1
   2
   2
   3
   3
   4
   4
   5
   5
(10 rows)

Time: 1777.162 ms (00:01.777)
materialize=> select * from philip limit 10; -- with `ALTER SYSTEM SET persist_fast_path_limit=0;`
 foo 
-----
   1
   1
   2
   2
   3
   3
   4
   4
   5
   5
(10 rows)

Time: 7112.857 ms (00:07.113)
```

This should be the pessimal case for this optimization, since it creates one or two large parts instead of many small ones like a more organic load pattern, but even so cutting out the other overheads is pretty significant.

Do you recall what you'd set `persist_fast_path_limit` to during your tests? Seeing the explain plans for the two queries might also be helpful.

File: test/testdrive/mzcompose.py
Comment (philip-stoev) [reply test/testdrive/mzcompose.py:30]: Historically we have had a lot of trouble keeping the settings used in the CI aligned with production defaults and/or picking the best settings to use for testing regardless of the production defaults. Hence the need to keep them all unified in `DEFAULT_SYSTEM_PARAMETERS` . So putting a custom default in testdrive/mzcompose.py would be a step backwards in that sense.

Consider either enabling this for the entire CI or augmenting your existing SLT file with a few SELECTs.

**Comment by philip-stoev** [comment]
I did this:

````
\timing on
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (f1 INTEGER);
SET statement_timeout='60s';
INSERT INTO t1 SELECT * FROM generate_series(1, 10 * 1000 * 1000);
EXPLAIN SELECT * FROM t1 LIMIT 999;
EXPLAIN SELECT * FROM t1 LIMIT 1000;
SELECT * FROM t1 LIMIT 999;
SELECT * FROM t1 LIMIT 1000;
```

The explains are correct:

````
materialize=> EXPLAIN SELECT * FROM t1 LIMIT 999;
            Optimized Plan             
---------------------------------------
 Explained Query (fast path):         +
   Finish limit=999 output=[#0]       +
     PeekPersist materialize.public.t1+
 
(1 row)

Time: 11.095 ms
materialize=> EXPLAIN SELECT * FROM t1 LIMIT 1000;
            Optimized Plan             
---------------------------------------
 Explained Query:                     +
   Finish limit=1000 output=[#0]      +
     ReadStorage materialize.public.t1+
 
```

and the performance is:

```
materialize=> SELECT * FROM t1 LIMIT 999;
Time: 3908.632 ms (00:03.909)
materialize=> SELECT * FROM t1 LIMIT 1000;
Time: 3586.826 ms (00:03.587)
```

Initially I thought I may be using unrealitic LIMIT sizes, but I also tried with 99-100 and got identical results (while setting `persist_fast_path_limit = 100`.

This is all with testdrive's mzcompose file:
```
cd test/testdrive
./mzcompose down -v ; ./mzcompose run default foo.td
```

and then I connect with
```
psql -p  `docker port testdrive-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
```

File: test/sqllogictest/persist-fast-path.slt
Comment (bkirwi) [reply test/sqllogictest/persist-fast-path.slt:152]: Offset was easy enough to support, so I've enabled it and added some tests.

**Comment by bkirwi** [comment]
I've added the tests that were straightforward to add, and queued up an issue for adding more and getting this enabled in CI: https://github.com/MaterializeInc/materialize/issues/22042

**Comment by bkirwi** [comment]
@philip-stoev - thank you for the script!

I just ran through some tests again: I still see decent speedups in staging and on bin/environmentd, but not when running in mzcompose. (And get timings comparable to what you saw.) I'm not sure why that is the case, though I'm glad I can reproduce it! That seems like a blocker to getting good CI tests for the feature, though I'd still like to get this into `main` and iterate from there if possible.

**Review by philip-stoev — APPROVED** [review]
@bkirwi please proceed as you see fit. In a separate PR, I am going to add a feature benchmark for this optimization to the feature benchmark (we are currently missing a LIMIT with no ORDER BY in the benchmark and it will be nice to have it in general).

File: src/adapter/src/coord/peek.rs
Comment (def-) [inline src/adapter/src/coord/peek.rs:167]: Could add an EXPLAIN test for this.

File: src/adapter/src/coord/peek.rs
Comment (def-) [inline src/adapter/src/coord/peek.rs:521]: Uncovered

File: src/adapter/src/coord/peek.rs
Comment (def-) [inline src/adapter/src/coord/peek.rs:557]: Testable by cancelling a long-running query?

File: src/persist-client/src/read.rs
Comment (def-) [inline src/persist-client/src/read.rs:1160]: Untested

File: src/persist-client/src/iter.rs
Comment (def-) [inline src/persist-client/src/iter.rs:493]: Untested

**Review by def- — COMMENTED** [review]
Some comments based on coverage report: https://buildkite.com/materialize/coverage/builds/239#018ae08a-7933-46cd-a0e5-92dde99d18b8

**Comment by bkirwi** [comment]
Thanks! I've linked your coverage report to our testing issue: https://github.com/MaterializeInc/materialize/issues/22042

As we build out more comprehensive testing for this (before anything gets enabled in prod, of course) we can make sure that those gaps are addressed. (Some of them may remain unreachable, and we can probably turn those into panics.)

**Pull request merged**
